DatadogCore/Sources/Core/Upload/DataUploadWorker.swift

```diff
@@ -55,95 +63,113 @@ internal class DataUploadWorker: DataUploadWorkerType {
         self.contextProvider = contextProvider
         self.backgroundTaskCoordinator = backgroundTaskCoordinator
         self.delay = delay
+        self.maxBatchesPerUpload = maxBatchesPerUpload
         self.featureName = featureName
         self.telemetry = telemetry
 
-        let uploadWork = DispatchWorkItem { [weak self] in
-            guard let self = self else {
+        scheduleNextCycle()
+    }
+
+    private func scheduleNextCycle() {
+        queue.asyncAfter(deadline: .now() + delay.current) { [weak self] in
+            guard let self = self, !self.cancelled  else {
                 return
             }
-
             let context = contextProvider.read()
-            let blockersForUpload = self.uploadConditions.blockersForUpload(with: context)
+            let blockersForUpload = uploadConditions.blockersForUpload(with: context)
             let isSystemReady = blockersForUpload.isEmpty
-            let nextBatch = isSystemReady ? self.fileReader.readNextBatch() : nil
-            if let batch = nextBatch {
+            let files = isSystemReady ? fileReader.readFiles(limit: maxBatchesPerUpload) : nil
+            if let files = files, !files.isEmpty {
+                DD.logger.debug("⏳ (\(self.featureName)) Uploading batches...")
                 self.backgroundTaskCoordinator?.beginBackgroundTask()
```

Comment:
not related to this PR but makes me think when `backgroundTaskCoordinator` be nil and what would happen in that case. IMO, we should do an early return.

Reply:
`backgroundTaskCoordinator` only registers tasks to get some background time after app goes into background. When it's `nil` we can just ignore this code and assume that this code can happen in foreground only.